### PR TITLE
ci(e2e): make a failing E2E run conclude failure instead of cancelled

### LINKED ISSUE #2206: ci: a failing E2E run on develop reports as cancelled, so the develop-red alarm never fires
## Symptom

A failing E2E run on `develop` reports to GitHub as **`cancelled`**, not `failure`.
`ci-health` deliberately excludes `cancelled` (conformance workflows set
`cancel-in-progress`, so a superseded run is benign), so the develop-red alarm
never fires and the failure is invisible. Five consecutive E2E runs on `develop`
lost their signal this way and nobody noticed.

## Evidence — run 33018102610 (attempt 2)

| when | what |
|---|---|
| 10:26:11 | job starts |
| 10:28:44 | `Run E2E tests` starts |
| 10:59:00 | `panic: test timed out after 30m0s` → `FAIL github.com/marmos91/dittofs/test/e2e 1800.031s` → `ok .../test/e2e/snapshot 0.706s` → bare `FAIL` |
| 10:59:00 → 11:41:23 | **42 minutes of no output at all** |
| 11:41:23 | `##[error]The operation was canceled.` — the job's `timeout-minutes: 75` fires |
| 11:41:28 | `Cleaning up orphan processes` / `Terminate orphan process: pid (11869) (tee)` |

Every other step in the job succeeded. Only `Run E2E tests` is `cancelled`, and a
cancelled job concludes `cancelled` regardless of what the other steps did.

Run 32979575910 has the identical signature: panic at 14:53:14, cancellation at
15:35:34 (42m later), single leaked orphan `tee`.

## What is actually wrong

The test step is a pipeline:

```
sudo -E env "PATH=$PATH" ... nix develop .#ci --command \
  go test -tags=e2e -v -timeout 30m ./test/e2e/... \
  2>&1 | tee "$GITHUB_WORKSPACE/e2e.log"
```

`go test` exited (it printed its final summary line). `tee` did not: a pipe reader
only sees EOF when **every** write end is closed, and the runner's cleanup confirms
`tee` was still alive 42 minutes later. So some process the suite left behind still
held the inherited descriptor. `bash` waits for `tee`; the step never ends; the job
burns its remaining budget; `timeout-minutes` converts a genuine test failure into
a cancellation.

Three separate defects stack here:

1. **The step can outlive the thing it runs.** Nothing bounds the pipeline, so a
   leaked descriptor holder consumes the whole job budget.
2. **The grading is unreachable.** `Generate summary` runs under `if: always()`
   and does grade the log — but a cancelled *job* concludes `cancelled` no matter
   what the summary step says. The step that runs the tests is the one that has
   to fail, and it never does when it is killed from outside.
3. **`-timeout 30m` inside `timeout-minutes: 75`** leaves ~42 minutes of
   unattributed slack between the suite's own deadline and the job's, which is
   exactly the window the wedge lived in.

## Not in scope here

The two test-level defects visible in the same run — `TestAdapterLifecycle/ADP-08`
and the `TestStoreMatrixV4` wedge that consumed the 30m test timeout — are tracked
separately. This issue is only about the workflow losing the failure signal.

## Fix shape

The test step must conclude `failure` for a run whose tests failed, panicked, or
never completed, and it must do so promptly enough that the job timeout is never
the thing that ends the run.

### PR DESCRIPTION
**Pull request by marmos91**
Closes #2206

## What was wrong

A failing E2E run on `develop` reported to GitHub as **`cancelled`**, not `failure`.
`ci-health.yml` deliberately excludes `cancelled` from the develop-red alarm — the
conformance workflows set `cancel-in-progress`, so a superseded run is benign — and
that exclusion swallowed genuine failures too.

Run 33018102610, attempt 2:

| when | what |
|---|---|
| 10:28:44 | `Run E2E tests` starts |
| 10:59:00 | `panic: test timed out after 30m0s` → `FAIL .../test/e2e 1800.031s` → `ok .../test/e2e/snapshot 0.706s` → bare `FAIL` |
| 10:59:00 → 11:41:23 | **42 minutes, no output at all** |
| 11:41:23 | `##[error]The operation was canceled.` — the job's `timeout-minutes: 75` fires |
| 11:41:28 | `Cleaning up orphan processes` / `Terminate orphan process: pid (11869) (tee)` |

Run 32979575910 has the identical signature: panic 14:53:14, cancellation 15:35:34
(42m later), single leaked orphan `tee`.

`go test` had exited — it printed its final summary line. `tee` had not, and the
runner's own cleanup says so 42 minutes later. A pipe reader sees EOF only when
**every** write end is closed, so something the run left behind still held the
inherited descriptor. `bash` waits for `tee`; the step never ends; the job burns
its remaining budget; `timeout-minutes` converts a test failure into a
cancellation — and a cancelled job concludes `cancelled` no matter what its other
steps did.

## Two premises checked, one refuted

**"A leaked `dfs` inherited the pipe write end" — refuted.** No Go-spawned child in
the suite is *given* the step's stdout or stderr. `grep -rn 'os.Stdout|os.Stderr'
test/e2e/` finds exactly one hit, and it is an `os.Stdout.Sync()`. Every process
the suite starts has its stdio pointed elsewhere:

- `test/e2e/helpers/server.go:419`, `:124` — `dfs start --foreground`, stdout/stderr to a log file
- `cmd/dfs/commands/daemon_unix.go:91` — the daemonising path, log file + `Setsid`
- `test/e2e/smb_byte_range_lock_test.go:63` — re-exec'd lock holder, `StdoutPipe()`, stderr `/dev/null`
- `test/e2e/framework/{netns,mount,kerberos}.go` — `CombinedOutput()` or nil, i.e. `/dev/null`
- `test/e2e/testdata/nlm/nlm_axis_interop.sh:80` — backgrounded server block redirected to a file

Go's `os/exec` hands a child `/dev/null` for any unset stream, so that specific
story does not hold. What the evidence *does* pin is the mechanism: the pipeline
could only end once every inherited write end was closed, one was not, and the step
outlived the thing it was running. The retained logs cannot name the holder — the
runner named only `tee`, because a root-owned descendant that has left the runner's
process tree is invisible to its orphan sweep. So the fix is structural rather than
aimed at one process, and a diagnostic is included so the next occurrence names its
own leaker.

**"`timeout` cannot bound a `sudo` child" — also refuted, and this one changed the
diff.** I built a signal-free watchdog on the theory that GNU `timeout` in front of
`sudo` would try to signal a privileged process, be refused, and then block in
`wait()` forever — reintroducing the same bug in a new shape. Measured on an
`ubuntu-latest` runner instead of reasoned about:

```
+ timeout --kill-after=3s 5s sudo -n sleep 120
##[error]Process completed with exit code 124.      # at 5s, not 8s, not never
```

`timeout` returns 124 at the wall, and at the wall rather than after the
`--kill-after` grace, so the TERM was delivered and `sudo` relayed it. The watchdog
was ~15 lines solving a problem that does not exist; it is gone, and the wall is
one `timeout`. The self-test keeps a `sudo_wedge` case so the assumption stays
measured rather than remembered.

## The fix

`.github/scripts/run-e2e.sh` owns the invocation and the grading.

1. **No pipe.** The suite's combined output goes straight to `$E2E_LOG`; the file is
   replayed to stdout afterwards, when reading it cannot block. A leaked holder of a
   file descriptor is merely leaked, not load-bearing.

2. **A wall.** `timeout --signal=TERM --kill-after=2m 45m` around the whole
   invocation, so a wedge anywhere below fails *this step* rather than outliving it.
   When it fires, the step dumps the process table — the holder of a wedge like this
   used to be invisible, which is why it took five silent runs to notice one.

3. **Grading in the step that runs the tests.** `Generate summary` already graded the
   log under `if: always()`, but that could never help: a cancelled job concludes
   `cancelled` whatever its steps say. Grading now happens where it can set the
   conclusion, on both the exit status and the log — status alone was already
   supposed to be sufficient (`set -o pipefail`) and still lost the signal, so a
   zero status with a log showing no completed pass is a failure.

### The numbers

| knob | value | why |
|---|---|---|
| `go test -timeout` | **30m**, unchanged | the suite's own budget; it fired exactly as designed in the failing run |
| `E2E_WALL` | **45m** | observed worst case is 30m16s of productive work (step 10:28:44, last output 10:59:00) — the 30m test timeout plus ~16s of cached dependency compile. 45m leaves ~15m of headroom, so a legitimately slow run finishes while a wedge fails ~15m past the suite's own deadline instead of ~45m past it |
| job `timeout-minutes` | **75**, unchanged | backstop only. Setup is ~2.5m, so the step now fails by ~48m and the job ends around 50m — the job timeout keeps ~25m of slack and is never the thing that fires. That is the whole point: it is the step that has to fail |

### By construction, on the runner

`.github/scripts/run-e2e-selftest.sh` drives the harness with a stand-in per way a
run can lose its signal, and is wired into `lint.yml` so it runs on every change.
From `ubuntu-latest`, run 33150884053:

```
ok   pass: exit 0 in 1s — passed
ok   fail: exit 1 in 2s — ::error title=E2E suite failed::test command exited 1
ok   leak: exit 1 in 1s — ::error title=E2E suite failed::test command exited 1
ok   wedge: exit 1 in 5s — ::error title=E2E suite failed::no result after 5s — the suite or something it started never finished
ok   deaf: exit 1 in 7s — ::error title=E2E suite failed::killed by SIGKILL after 7s (wall 5s + 2s) — the wall escalating past TERM, or the out-of-memory killer
ok   green: exit 1 in 1s — ::error title=E2E suite failed::no package reported ok — the suite did not run to completion
ok   errexit: exit 1 in 0s — graded despite inherited errexit
ok   sudo_wedge: exit 1 in 5s — ::error title=E2E suite failed::no result after 5s — the suite or something it started never finished
ok   control: old pipeline shape still running after 10s on the same stand-in
all cases passed
```

- `leak` is the reported defect: the stand-in prints a timeout panic, exits, and
  leaves a child holding the inherited descriptors.
- `control` runs the **old** `... 2>&1 | tee log` shape against that same stand-in
  and asserts it does *not* finish. Without it the `leak` case would be vacuous.
- `wedge` never finishes and must fail at the wall; `sudo_wedge` is the same under
  `sudo`, which is where the wall's signalling assumption actually lives.
- `deaf` ignores `TERM`, forcing the wall to escalate to `SIGKILL`. That is a
  different exit status and a different diagnosis, because 137 is any `SIGKILL` —
  measured: `timeout --kill-after=1s 2s bash -c 'trap "" TERM; sleep 30'` and
  `timeout --kill-after=5s 60s bash -c 'kill -9 $$'` both give 137, while a plain
  timeout gives 124. The status alone cannot separate the wall from the
  out-of-memory killer, so the message reports it as `SIGKILL` with the elapsed
  time, which can.
- `green` exits 0 having produced no result.
- `errexit` invokes the harness through `env SHELLOPTS=errexit`. Bash does import an
  exported `SHELLOPTS`, and errexit would abort the harness at the test command,
  before it reads the exit status grading depends on. It cannot arrive that way from
  a workflow step, but stripping the `set +e` and running the self-test under an
  inherited `-e` reduces it from eight assertions to one, which is why both scripts
  clear it explicitly.
- Each case asserts the *reason* it failed, not just the exit code. That caught a
  real self-satisfying test: a stand-in rewritten to use `exec -a` died instantly
  because `sleep` here is a multi-call coreutils binary that dispatches on `argv[0]`,
  and the wedge case sailed through on the right exit code for entirely the wrong
  reason.

### The guards were tested against deliberately broken harnesses

Each load-bearing property was removed in turn and the self-test re-run. Every one
is refused, by the case that targets it:

| harness broken by | self-test result |
|---|---|
| restoring the `\| tee` pipeline | `FAIL leak: harness did not return within 20s`, `FAIL wedge: harness did not return within 25s`, `FAIL wedge: failed for the wrong reason` |
| deleting the log grading | `FAIL green: exit 0, want 1`, `FAIL green: failed for the wrong reason` |
| deleting the wall | `FAIL wedge: harness did not return within 25s`, `FAIL wedge: failed for the wrong reason` |
| deleting the `SIGKILL` branch | `FAIL deaf: failed for the wrong reason (wanted "killed by SIGKILL")` |
| deleting the harness's `set +e` | `FAIL errexit: exit 1, and grading did not run` |
| deleting the self-test's `set +e`, under inherited `-e` | one case reported instead of eight, exit 1 |

### End to end on the real suite

`workflow_dispatch` on this branch, run
[33151205489](https://github.com/marmos91/dittofs/actions/runs/33151205489). The suite
is genuinely broken right now (#2207, #2208, #2209), so this is the same failure that
reported `cancelled` on `develop`, run through the new path:

| | before (33018102610) | after (33151205489) |
|---|---|---|
| `Run E2E tests` step | **cancelled** | **failure** |
| job conclusion | **cancelled** | **failure** |
| step duration | 72m54s — 30m16s of work, then 42m of nothing | **30m18s** — ends when `go test` ends |
| job duration | 75m19s, killed by `timeout-minutes` | 33m22s, ended by its own verdict |
| develop-red alarm | skipped (`cancelled` is excluded) | fires |

The 42-minute post-`FAIL` hang is gone: the step now ends 18 seconds after `go test`
hits its 30m timeout, instead of 42 minutes after. Every other step still succeeded,
exactly as before — the difference is that the step running the tests is now the one
that fails, so the job conclusion follows it.

## Sibling exposure, deliberately not fixed here

`posix-tests.yml:214` has the same `... 2>&1 | tee "$LOG_FILE" || true` shape and the
same latent exposure. It has not hit it: all its recent `cancelled` runs on `develop`
last ~30 seconds, i.e. ordinary `cancel-in-progress` supersedes, not wedges. Left
alone rather than speculatively rewritten.

Worth knowing when reading run history: `cancelled` on these workflows is genuinely
ambiguous today. Of the recent E2E cancellations on `develop`, two (33018102610,
32979575910) ran ~75m and are wedges; the rest lasted seconds to minutes and are
benign supersedes. The alarm cannot tell them apart, which is precisely why the
wedge hid for five runs.

### CODE REVIEW

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -120,3 +120,30 @@ jobs:
...
+          set -x
+          sudo -n true && echo "passwordless sudo: yes"
+          start=$SECONDS
+          timeout --kill-after=3s 5s sudo -n sleep 120 >/dev/null 2>&1
+          echo "timeout-outside-sudo: exit $? after $((SECONDS - start))s"
+          start=$SECONDS
+          sudo -n timeout --kill-after=3s 5s sleep 120 >/dev/null 2>&1
+          echo "timeout-inside-sudo:  exit $? after $((SECONDS - start))s"
+          ps -eo pid,user,etimes,args | grep -E "sleep 120" | grep -v grep || echo "no sleep 120 survivors"
+          sudo -n pkill -f "sleep 120" || true
```
[the remark this answers is not in the document]
Comment (marmos91) [reply .github/workflows/lint.yml]: Gone. That step was a temporary on-runner probe to settle whether \`timeout\` can bound a \`sudo\` child; it has served its purpose (it can — exit 124 at exactly 5s, in run 33150884053) and the branch has been force-pushed without it. Your reading of it was right, though: it did abort the step on the expected non-zero, which is how the answer reached the log at all.

File: .github/workflows/e2e-tests.yml [hunk 2]
```diff
@@ -118,21 +119,29 @@ jobs:
...
 
       - name: Run E2E tests
+        env:
+          E2E_LOG: ${{ github.workspace }}/e2e.log
+          # 30m of test timeout plus room for the sudo-side dependency
+          # fetch+compile, the panic dump, and the remaining packages. The
+          # observed worst case is 30m16s of productive work, so this fails a
+          # wedge roughly 15m past a legitimately slow run rather than letting
+          # it run out the job.
+          E2E_WALL: 45m
```
[the remark this answers is not in the document]
Comment (marmos91) [reply .github/workflows/e2e-tests.yml:129]: Fixed by removing the cause rather than the symptom. \`E2E_WALL_SECONDS\` belonged to a signal-free watchdog I built on the premise that \`timeout\` in front of \`sudo\` could not bound anything. Measured on a runner, that premise is false, so the watchdog is gone and the wall is one \`timeout\` again. The harness now reads \`E2E_WALL\` as a duration string, matching what the workflow exports.

File: .github/workflows/e2e-tests.yml [hunk 3]
```diff
@@ -45,10 +45,11 @@ jobs:
...
     runs-on: ubuntu-latest
-    # The suite runs `go test -timeout 30m`, plus a cold fetch+compile of the
-    # e2e-only test deps under sudo. 20m was only ever enough because the job
-    # was a no-op (go test never ran); a real run needs headroom past the 30m
-    # test timeout itself.
+    # Backstop only. The test step bounds itself at E2E_WALL (45m) so that a
+    # wedge fails the step; this number just has to stay comfortably above
+    # setup + that wall + the grace period, because a job killed by its own
+    # timeout concludes `cancelled` and every alarm that keys on `failure`
+    # stays silent.
```
[the remark this answers is not in the document]
Comment (marmos91) [reply .github/workflows/e2e-tests.yml:52]: Same root cause; the comment and the harness now both say \`E2E_WALL\` in duration units.

File: .github/scripts/run-e2e.sh [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+# delivering a signal it may not be permitted to deliver is not a wall. What the
+# wall must guarantee is that this script returns, and fails, while the enclosing
+# job still has budget — a job ended by its own timeout concludes `cancelled`
+# rather than `failure`, and alarms that key on `failure` then stay silent.
+#
+# Grading uses both the exit status and the log: the status alone was already
+# supposed to be sufficient and still lost the signal once, so a zero status
+# with a log that shows no completed pass is treated as a failure.
+set -uo pipefail
+
```
[the remark this answers is not in the document]
Comment (marmos91) [reply .github/scripts/run-e2e.sh]: Real, and taken — thank you. Bash does import \`SHELLOPTS\`, which I verified: \`env SHELLOPTS=errexit\` on a script whose own \`set\` line omits \`-e\` aborts it at the first non-zero command. It cannot reach the harness from the workflow step (Actions passes \`-e\` on the \`bash\` command line and does not export \`SHELLOPTS\`), but the consequence would be that the harness dies before reading the exit status grading depends on, and this is the code that decides whether a failure is visible at all. Both scripts now clear it explicitly, and the self-test has an \`errexit\` case that invokes the harness through \`env SHELLOPTS=errexit\` and asserts it still grades. Removing the \`set +e\` makes that case fail, so the guard is verified rather than merely present.

File: .github/scripts/run-e2e-selftest.sh [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+#
+# Self-test for run-e2e.sh. Drives the harness with stand-in test commands that
+# reproduce each way a real run loses its signal, and asserts the harness reports
+# failure, promptly, every time.
+#
+# Case 3 also runs the *old* pipeline shape against the same stand-in and asserts
+# it does NOT finish — without that, the case would pass against a harness that
+# never had to refuse anything.
+set -uo pipefail
+
```
[the remark this answers is not in the document]
Comment (marmos91) [reply .github/scripts/run-e2e-selftest.sh]: Same fix, same case — the self-test clears errexit explicitly too.

File: .github/scripts/run-e2e.sh [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+timeout --signal=TERM --kill-after="$GRACE" "$WALL" "$@" >"$LOG" 2>&1
+status=$?
+
+if [ "$status" -eq 124 ] || [ "$status" -eq 137 ]; then
+    # Name what was still running. The holder of a wedge like this used to be
+    # invisible, which is why it took five silent runs to notice one.
+    echo "::group::processes still running at the wall"
+    ps -eo pid,ppid,pgid,stat,etimes,user,args 2>/dev/null || true
+    echo "::endgroup::"
+fi
```
[the remark this answers is not in the document]
Comment (marmos91) [reply .github/scripts/run-e2e.sh:53]: Real, and measured rather than reasoned about. On this machine:

    timeout --signal=TERM --kill-after=1s 2s bash -c 'trap "" TERM; sleep 30'   -> 137
    timeout --signal=TERM --kill-after=5s 60s bash -c 'kill -9 $$'              -> 137
    timeout --signal=TERM --kill-after=5s 2s sleep 30                          -> 124

So 137 covers both the wall escalating past an ignored TERM and an external kill, and the status alone cannot separate them — you are right that calling it "the wall" is a guess.

I did not take the suggested shape of letting 137 fall through to the generic path, because the wall genuinely can produce it and the generic message ("test command exited 137") throws away the one thing that makes it diagnosable. Instead 137 now has its own branch that says what it is and prints the elapsed time, which is the discriminator: a wall kill lands at wall + grace, an out-of-memory kill lands wherever the allocation did. The process-table group is retitled "processes still running when the run was killed" and still fires for both, since it is useful in either case.

The branch is now exercised, not assumed — the self-test has a `deaf` stand-in that ignores TERM, forcing the escalation:

    ok   deaf: exit 1 in 7s — killed by SIGKILL after 7s (wall 5s + 2s) — the wall escalating past TERM, or the out-of-memory killer

Deleting the 137 branch makes that case fail with "failed for the wrong reason", so it is load-bearing.

On the smbtorture harness precedent: noted, and the difference is deliberate — it has no `--kill-after`, so 137 there really cannot come from its own wall.

File: .github/scripts/run-e2e.sh [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+cat "$LOG"
+
+# Anything of ours still running is a leak. It can no longer hold this step open,
+# but it is worth naming.
+survivors=$(pgrep -a -f '(^|/)(dfs|dfsctl|mount\.nfs|mount\.cifs|rpc\.statd|rpcbind)( |$)' 2>/dev/null)
+if [ -n "$survivors" ]; then
+    echo "::warning title=Leaked processes::the suite left processes running"
+    printf '%s\n' "$survivors"
+fi
```
[the remark this answers is not in the document]
Comment (marmos91) [reply .github/scripts/run-e2e.sh:66]: Taken as written. The workflow itself installs and starts rpcbind (`apt-get install -y ... rpcbind`) and `rpc-gssd`, so those two matched on every healthy run — a warning that always fires is read as background noise, which defeats the point of having it. The pattern is now `dfs|dfsctl|mount\.nfs|mount\.cifs`, i.e. only processes the suite is expected to own, with a comment recording why the daemons are excluded.

File: .github/scripts/run-e2e-selftest.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (marmos91) [reply .github/scripts/run-e2e-selftest.sh]: Settled by running it rather than reading it, because you are right that this one decides whether the rest of the evidence means anything.

First, the premise is real. Bash does import an exported \`SHELLOPTS\`:

    $ env SHELLOPTS=errexit bash -c 'echo "opts=$-"; false; echo STILL ALIVE'
    opts=ehBc            # note the e
    (no STILL ALIVE, exit 1)

Then the self-test itself, under an inherited \`-e\`:

    # with its own `set +e` (as merged)
    $ env SHELLOPTS=errexit ./.github/scripts/run-e2e-selftest.sh
    ok   pass ... ok fail ... ok leak ... ok wedge ... ok deaf ... ok green ... ok errexit ... ok control
    all cases passed                                            exit=0

    # with that `set +e` removed
    $ env SHELLOPTS=errexit ./.github/scripts/run-e2e-selftest.sh
    ok   pass: exit 0 in 0s — passed
    exit=1

So the failure mode is exactly what you described: it dies at the first case that exits non-zero, having asserted one of eight things. It exits 1 rather than 0, so it would not have reported a false success — but it would have reported a seventh of the evidence, which is the part that matters. Both scripts now clear errexit explicitly, and the self-test carries an `errexit` case that invokes the harness through `env SHELLOPTS=errexit` and asserts it still grades; deleting the harness's `set +e` makes that case fail, so the guard is verified rather than merely present.

For completeness on reachability: it cannot arrive this way from the workflow. Actions runs steps as `bash -e {0}`, i.e. `-e` on the command line, and does not export `SHELLOPTS`, so a child script does not inherit it. The guard is there because this is the code that decides whether a failure is visible at all, and it costs two characters.
